### DISCUSSION TITLE
Use fuzzy search algorithm like levenshtein to enhance current search?

### DISCUSSION THREAD

**Discussion by pc223** [Flow-Launcher/Flow.Launcher#597] [Ideas]
Does the current fuzzy match use levenshtein? I skimmed the code seems doesn't. 

The current algo seems not accept typo: "chrome" and "chorme"

Should we use something like this to do that?

https://github.com/JakeBayer/FuzzySharp

**Comment by taooceros** [comment]
That's a bit hard to use due to Flow requires the fuzzy to provide highlight data. We need to find a workaround for that when using levenshtein.

**Reply by pc223** [reply]
Do you able to import the FuzzySharp lib to Flow?

I want to experiment with it to compare it to current algorithm, but seems FuzzySharp not work with .net 5

**Reply by taooceros** [reply]
Do you have the error message? I think it should support net 5.

I am afk now, so need to test it later.

**Reply by pc223** [reply]
> Do you have the error message? I think it should support net 5.
> 
> I am afk now, so need to test it later. 

:+1: I'll update error message later

**Reply by taooceros** [reply]
import it with nuget should be fine.

**Reply by pc223** [reply]
Seems install thru nuget works fine now, not sure why I fail earlier😅

**Comment by pc223** [comment]
Demo advantages of levenshtein to current:

Try out at build of PR https://github.com/Flow-Launcher/Flow.Launcher/pull/616

Old behavior is left, new is right.
<p align="middle">
<img src="https://user-images.githubusercontent.com/10551242/127423662-8c44b0a8-9beb-405b-84e7-b319532863a0.gif" width=350>
<img src="https://user-images.githubusercontent.com/10551242/127423659-ac42ac9a-73f6-402d-937b-a4ab17afa58a.gif" width=350>
</p>

**Reply by jjw24** [reply]
just thought i put the options here what you can do with Fuzzy Search:

**Simple Ratio**
Fuzz.Ratio("mysmilarstring","myawfullysimilarstirng")
72

Fuzz.Ratio("mysmilarstring","mysimilarstring")
97

**Partial Ratio**
Fuzz.PartialRatio("similar", "somewhresimlrbetweenthisstring")
71

**Token Sort Ratio**
Fuzz.TokenSortRatio("order words out of","  words out of order")
100

Fuzz.PartialTokenSortRatio("order words out of","  words out of order")
100

**Token Set Ratio**
Fuzz.TokenSetRatio("fuzzy was a bear", "fuzzy fuzzy fuzzy bear")
100

Fuzz.PartialTokenSetRatio("fuzzy was a bear", "fuzzy fuzzy fuzzy bear")
100

Token Initialism Ratio
Fuzz.TokenInitialismRatio("NASA", "National Aeronautics and Space Administration");
89

Fuzz.TokenInitialismRatio("NASA", "National Aeronautics Space Administration");
100

Fuzz.TokenInitialismRatio("NASA", "National Aeronautics Space Administration, Kennedy Space Center, Cape Canaveral, Florida 32899");
53

Fuzz.PartialTokenInitialismRatio("NASA", "National Aeronautics Space Administration, Kennedy Space Center, Cape Canaveral, Florida 32899");
100

**Token Abbreviation Ratio**
Fuzz.TokenAbbreviationRatio("bl 420", "Baseline section 420", PreprocessMode.Full);
40

Fuzz.PartialTokenAbbreviationRatio("bl 420", "Baseline section 420", PreprocessMode.Full);
50      

**Weighted Ratio**
Fuzz.WeightedRatio("The quick brown fox jimps ofver the small lazy dog", "the quick brown fox jumps over the small lazy dog")
95

**Reply by pc223** [reply]
Yeah, a lot of interesting functions, can open up new features, in the demo I only use the `PartialTokenSetRatio`

**Comment by jjw24** [comment]
I doubt we can integrate it into our own search algo, might have to do it as a separate option, if so we also need to find out how long it takes to use this additional search so our search times are not taking longer.

**Reply by pc223** [reply]
Yeah, we can use this algo as a "backup" for if normal algo fails (or the score too low)

**Reply by jjw24** [reply]
something like an option in the settings tab where you switch on if you would like search to consider 'misspelt' words etc?

**Reply by pc223** [reply]
> something like an option in the settings tab where you switch on if you would like search to consider 'misspelt' words etc?

I think we can do that for now, to see if people like it, or does it have any performance problem. After that we should consider make that default and delete the option from settings tab if no need for the old behavior anymore.

**Reply by momadacoding** [reply]
Please do it

**Reply by jjw24** [reply]
Hi @haolly , are you having any issues with the current search method?

**Reply by momadacoding** [reply]
I can't search use partial name path, only use the full path name can show the result.
**Full name search result.**
![image](https://user-images.githubusercontent.com/2383231/158928946-5dd571b6-432c-4e1e-9919-87e1b0df20ac.png)

**Partial name search result**
![image](https://user-images.githubusercontent.com/2383231/158929067-0e40c7ed-d44d-4381-b1ee-d73e81b9dc17.png)

**This is the result by Listary**
![image](https://user-images.githubusercontent.com/2383231/158929168-12c03aca-40ac-43fc-b0d0-ce69af2ada43.png)
